var/spack/repos/builtin/packages/mathematica/package.py

```diff
@@ -30,10 +30,11 @@ def install(self, spec, prefix):
            '-execdir={0}'.format(prefix.bin),
            '-selinux=y')
 
-        # after the install phase completes, spack tries to install build.out
-        # into <prefix>/.spack, but the .spack dir will not exist, causing the
-        # build to fail. package.py:1690 seems to show that the dir is created
-        # right before writing build.out -- possible bug?
+        # After the install phase completes, Spack tries to install
+        # spack-build-out.txt into <prefix>/.spack, but the .spack dir will not
+        # exist, causing the build to fail. package.py:1690 seems to show
```

Comment:
(not that you wrote this, but...) This comment is likely not very useful anymore, since I think package.py has changed since it was written. Would you be able to ensure that this `.spack` directory exists (which would remove the need for this comment altogether)?

Reply:
The method `create_install_directory` (`YamlDirectoryLayout`) creates the subdirectory with the same permissions as the `prefix`.  There is nearly identical code in`package.py`'s `do_install` for ensuring the directory has the right group and permissions if it exists.

So I'm thinking about adding a method, say `ensure_install_layout` to `DirectoryLayout` (and `YamlDirectoryLayout`), that makes sure the directories (including `.spack` for the latter), group, and permissions are correct.

Questions:
- Should `YamlDirectoryLayout`'s `check_installed` be checking for `.spack`?
- If so, what should it do if `.spack` is not present (e.g., raise an exception)?

Reply:
> what should it do if .spack is not present (e.g., raise an exception)?

In this particular case I definitely think it should create the directory if it doesn't exist (don't raise an error).

> Should YamlDirectoryLayout's check_installed be checking for .spack

IMO no: that function is intended to check whether a package was installed at all, vs. whether it was installed correctly. I think if we create the `.spack` directory in `Package.log`, that would be fine.

Reply:
`Package`'s `do_install` will call `create_install_directory` -- so create the `.spack` subdirectory -- _unless_ the `prefix` already exists.  In the later case it takes the time to make sure the group and directory permissions are correct but does not ensure `.spack` exists.  IMHO that is what needs to change (in the current code base) since all of this is done before forking the build, which will call the `log` method to move the files from the build to the install directory.